sematic/plugins/external_resource/ray/cluster.py

```diff
@@ -0,0 +1,387 @@
+                    message,
+                ),
+                _cluster_name=cluster_name,
+            )
+
+        return replace(
+            self._with_status(
+                ResourceState.ACTIVATING,
+                f"Requested Ray cluster with name '{cluster_name}'.",
+            ),
+            _cluster_name=cluster_name,
+            _head_uri=head_uri,
+        )
+
+    @classmethod
+    def _k8s_client(cls) -> kubernetes.client.ApiClient:  # type: ignore
+        load_kube_config()
+        api_client = kubernetes.client.ApiClient()  # type: ignore
+        return api_client
+
+    @classmethod
+    def _cluster_api(cls) -> kubernetes.dynamic.DynamicClient:  # type: ignore
+        load_kube_config()
+        api = kubernetes.dynamic.DynamicClient(  # type: ignore
+            cls._k8s_client()
+        ).resources.get(api_version="ray.io/v1alpha1", kind="RayCluster")
+        return api
+
+    @classmethod
+    def _get_kuberay_version(
+        cls, namespace: str
+    ) -> Tuple[Optional[str], Optional[str]]:
+        try:
+            api_instance = kubernetes.client.AppsV1Api(cls._k8s_client())
+            name = cls._kuberay_wrapper().KUBERAY_DEPLOYMENT_NAME
+            api_response = api_instance.read_namespaced_deployment(name, namespace)
+            ready: Optional[int] = api_response.status.ready_replicas  # type: ignore
+
+            if ready is None:
+                ready = 0
+
+            if ready < 1:
+                message = (
+                    "Kuberay has no ready replicas. Please ask your "
+                    "cluster administrator to verify the health of Kuberay, "
+                    "and refer to Kuberay docs for troubleshooting: "
+                    "https://ray-project.github.io/kuberay/"
+                )
+                logger.error("Kuberay is not healthy. Status: %s", api_response.status)
+                logger.error("Kuberay is not healthy. Message: %s", message)
+                return (None, message)
+            for container in api_response.spec.template.spec.containers:  # type: ignore
+                if container.name == cls._kuberay_wrapper().KUBERAY_CONTAINER_NAME:
+                    image_tag = container.image.split(":")[-1]  # type: ignore
+                    return image_tag, None
+        except Exception as e:
+            logger.exception("Error getting Kuberay version: %s", e)
+            return None, str(e)
+
+        return (
+            None,
+            "Kuberay version could not be determined from Kuberay deployment.",
+        )
+
+    def _do_update(self) -> "RayCluster":
+        if self.status.state == ResourceState.ACTIVATING:
+            return self._update_from_activating()
+        elif self.status.state == ResourceState.ACTIVE:
+            return self._update_from_active()
+        elif self.status.state == ResourceState.DEACTIVATING:
+            return self._continue_deactivation("Continuing deactivation")
```

Comment:
Is it actually needed to send the signal again?

Reply:
`_continue_deactivation` works by sending a `delete` for the cluster resource. If the cluster resource existed, then that's all it does. If the cluster does not exist, the `delete` gets a 404, and that's the indicator that the cluster has been fully deactivated.